### DISCUSSION TITLE
Testing Redirect with headers

### DISCUSSION THREAD

**Discussion by mojowill** [livewire/livewire#4218] [1. Help]
I have a general use component for deleting models, once the delete has happened I redirect using:

```php
return redirect(request()->header('Referer'));
```

I can't hardcode the redirect as it may be used in multiple places and I need to trigger a full page reload.

Now I'm trying to test this component but can't seem to find any valid assertions to use.

The most obvious one was to assertRedirect() however as the header doesn't exist the assertion fails with:

```
Component did not perform a redirect.
Failed asserting that an array has the key 'redirect'.
```

My test looks like this:

```php
Livewire::test(DeleteConfirmation::class, ['model' => $folder])
            ->withHeaders(['Referer' => route('team.folders.index')])
            ->call('delete')
            ->assertRedirect(route('team.folders.index'));
```

Can anyone give me a pointer? Is there better way to redirect or refresh the current page?

**Comment by joshhanley** [comment]
@mojowill I'm not sure, if `withHeaders` doesn't work, maybe you will have to dig into the `TestableLivewire` class that the `Livewire::test` helper returns and see if you can interact with it to set the headers some other way.

**Reply by mojowill** [reply]
Thanks for your reply I was able to solve this by refactoring the component to allow for the absence of the header which meant in the test environment it would work.

I'd be curious to know if there is a better way of doing the redirect? Essentially I just want to go back to the previous page with a reload (the component needs to force a page reload as the parent is not a Livewire component in all cases).

**Reply by joshhanley** [reply]
@mojowill no worries!

Hmm I think what I would do, is in the mount method of your component, I would store the current route in a public property (as when Livewire makes a request, it says the current route is Livewire's `/livewire/messages` endpoint).

Then in your method just do a redirect to the stored route.

Or change and use a full page Livewire component, then you have no need for refreshing the whole page (this is what I would do) 😁

Hope this helps!

**Comment by ItsANameToo** [comment]
One option to keep the redirect as-is is by using Laravel's `previous` url instead of relying on the `Referer` header directly.

```
return redirect(url()->previous(
    fallback: route('optional-fallback-route')
));
 ```
 
 This will allow you to set the previous url by using a session, as the above code will check for a `refer` header or a `_previous.url` session variable to redirect.
 
 In a test you can then add this to make it work with the redirect
 
 ```
 $this->withSession([
    '_previous' => [
        'url' => '/your-route'
    ],
]);
  ```
